modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlledQueueMessageDeliveryImpl.java

```diff
@@ -103,55 +112,58 @@ public int deliverMessageToSubscriptions(StorageQueue storageQueue) throws Andes
                         }
 
                         message.markAsScheduledToDeliver(localSubscription);
+                        //if the message was delivered, it needs to be removed
                         iterator.remove();
                         MessageFlusher.getInstance().deliverMessageAsynchronously(localSubscription, message);
                         numOfCurrentMsgDeliverySchedules++;
 
                         //for queue messages and durable topic messages (as they are now queue messages)
                         // we only send to one selected subscriber if it is a queue message
-                        break;
-                    }
-                }
-
-                //if the message was delivered, it needs to be removed
-                if (numOfCurrentMsgDeliverySchedules == 1) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Removing Scheduled to send message from buffer. MsgId= " + message.getMessageID());
-                    }
-                    sentMessageCount++;
-                } else {
-
-                    //if no subscriber has a matching selector, route message to DLC queue
-                    if (!subscriberWithMatchingSelectorFound && !suspendedSubFound) {
```

Comment:
In a way Yes. We use a minCursor that is the minimum cursor of all the subscribers. Till another subscription comes all the values in front of this minCursor in the buffer would be removed but would remain in the db.